### PR TITLE
Add state-management service

### PR DESCRIPTION
**Pull request by kuzmavictor**
During this iteration was implemented:

- feature of the state-management system(located on the following path: `name_application\service\store` );

- components for file item and file-list (`FileList`, `File-Item`);

- processes of fetching file list and removing an item from file list using implemented state-management system.

### CODE REVIEW

**Comment by kuzmavictor** [comment]
@dkalinin PTAL.

File: .gitignore [hunk 1]
```diff
@@ -1,75 +1,75 @@
...
-com_crashlytics_export_strings.xml
-crashlytics.properties
-crashlytics-build.properties
-fabric.properties
-
-
-Status API Training Shop Blog About
-© 2016 GitHub, Inc. Terms Privacy Security Contact Help
-
+*.class
```
Comment (dkalinin) [inline .gitignore]: Please remove all items that are not used in this project.

File: app/controller/file-list/index.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    super( parentRoot, router );
+    this._fileListPage = new FileListPage( this._root );
+    this._creteFileListPage();
+  }
+
+  /**
+   *
+   * @private
+   */
+  _creteFileListPage() {
```
Comment (dkalinin) [inline app/controller/file-list/index.js]: cre-a-te

File: app/service/store/action/index.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import {FileListService} from '../../request/file-list';
+import {state as states} from '../state';
```
Comment (dkalinin) [inline app/service/store/action/index.js]: This import is not needed. You have `context` param in actions.

File: app/service/store/pub-sub/index.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  /**
+   * Provides functionality of publishing event(returns the callbacks for an event)
+   *
+   * @param {String} event - a name of event
+   * @param {Object} data - properties with which subscribed listeners will be called
+   * @return {Array} - the callback for current event
+   */
+  publish(event, data) {
+    return this.events[event].map((callback) => callback(data));
```
Comment (dkalinin) [inline app/service/store/pub-sub/index.js]: Check if this.events[event] is defined.

File: app/view/component/button/index.js [hunk 5]
```diff
@@ -38,6 +38,14 @@ export class Button extends Component {
     this._root.appendChild( this.button );
   }
 
+  /**
+   * Sets disable attribute.
```
Comment (dkalinin) [inline app/view/component/button/index.js]: disabled

File: app/view/component/file-list/index.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        fileItem.onClick(() => {
+          fileItem.disabledButtonElement = true;
+          this.dispatch('removeItem', element.id);
+        });
+      });
+    }
+  }
+
+  /**
+   * Creates span element that provide rendering process of loading file-list
```
Comment (dkalinin) [inline app/view/component/file-list/index.js]: provide-s

File: app/view/component/file-list/index.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    } else {
+      this._root.innerHTML = '';
+      const ul = document.createElement('ul');
+      this._root.appendChild(ul);
+
+      this.state.files.forEach((element) => {
+        const li = document.createElement('li');
+        ul.appendChild(li);
+
+        const fileItem = new FileItem(li, element.name, element.id);
```
Comment (dkalinin) [inline app/view/component/file-list/index.js]: `li` can be a root element of FileItem.

File: app/application/index.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  /**
+   * Configures process initializing controllers and
+   * service routing {@link Router} for application.
+   *
+   * @see AuthorizationController
+   * @see RegistrationController
+   * @see Page404Controller
+   * @private
+   */
+  _render() {
```
Comment (dkalinin) [inline app/application/index.js]: This method configures Router instance.

File: app/application/index.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+   * @see Store
+   * @see RequestServiceProvider
+   * @private
+   */
+  _configController() {
+    const requestServiceProvider = new RequestServiceProvider();
+    const requestServices = requestServiceProvider.requestServices;
+    console.log(requestServices);
+    const store = new Store(requestServices);
+    return {requestServices, store};
```
Comment (dkalinin) [inline app/application/index.js]: Please avoid returning hashes in order to pass two variables.

File: app/controller/file-list/index.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import {BaseController} from '../';
+import {FileListPage} from '../../view/component/page/file-list';
+
+/**
+ * This class provides controller-layer for file-list
```
Comment (dkalinin) [inline app/controller/file-list/index.js]: ...file list.

File: app/service/request/index.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+   * This constructor initializes {@link RequestServiceProvider}
+   *
+   * @see AuthorizationService
+   * @see RegistrationService
+   * @see FileListService
+   */
+  constructor() {
+    this._authorizationService = new AuthorizationService();
+    this._registrationService = new RegistrationService();
+    this._fileListService = new FileListService();
```
Comment (dkalinin) [inline app/service/request/index.js]: These can be local variables.

File: app/service/store/action/index.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+ * Action implement all business logic for components that use store (state-management system).
+ */
+export const actions = {
+
+  /**
+   * Removes an item {@link FileItem} from file list {@link FileList}.
+   * The promise that returns emulate working with server-side.
+   *
+   * @param {Store} store - an instance of {@link Store}
+   * @return {Function}
```
Comment (dkalinin) [inline app/service/store/action/index.js]: `@return {function(int)} - `

File: app/service/store/index.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+ * This class implement the application class of state-management system inside this application.
+ * It contains state object, commits methods that call mutations,
+ * and dispatch method that calls actions. Inside  of this store object is the Proxy object,
+ * that monitors the state of object, using the functionality of PubSub.
+ */
+export class Store {
+  /**
+   * The constructor of {@link Store}. It initializes the store.
+   *
+   * @param {Object} requestServices - an object that contains instances all request services
```
Comment (dkalinin) [inline app/service/store/index.js]: Let's pass service provider here.

File: app/service/store/index.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  /**
+   * Calls an action if it exists.
+   *
+   * @param {String} name - a name of action
+   * @param {...*} params - params  that contains specified property for current action
+   */
+  dispatch(name, ...params) {
+    this.actions[name](this)( ...params);
```
Comment (dkalinin) [inline app/service/store/index.js]: this.actions[name] can be undefined.

File: app/service/store/index.js [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  /**
+   * Calls a mutator if it exists.
+   *
+   * @param {String} name - a name of mutation
+   * @param {...*} params -  an object that contains specified property for current mutator
+   */
+  commit(name, ...params) {
+    this.mutations[name](this.state, ...params);
```
Comment (dkalinin) [inline app/service/store/index.js]: this.mutations[name] can be undefined.

File: app/view/component/file-item/index.js [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    this._render();
+  }
+
+  /**
+   * This method allows to render file item all its own fields.
+   *
+   * @private
+   */
+  _render() {
+    this.htmlliElement = document.createElement(LI_TAG);
```
Comment (dkalinin) [inline app/view/component/file-item/index.js]: htmlliElement => htmlLiElement

File: app/view/component/file-item/index.js [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    this.htmlDivElement = document.createElement(DIV_TAG);
+    this.fileNameSpan = new Span(this.htmlDivElement);
+    this.fileNameSpan.text = this._name;
+    this.button = new Button(this.htmlDivElement, BUTTON_TEXT);
```
Comment (dkalinin) [inline app/view/component/file-item/index.js]: button field should be private.

File: app/view/component/file-item/index.js [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+  /**
+   * This method allows to render file item all its own fields.
+   *
+   * @private
+   */
+  _render() {
+    this.htmlliElement = document.createElement(LI_TAG);
+    this.htmlDivElement = document.createElement(DIV_TAG);
+    this.fileNameSpan = new Span(this.htmlDivElement);
+    this.fileNameSpan.text = this._name;
```
Comment (dkalinin) [inline app/view/component/file-item/index.js]: Please remove unused fields.

File: app/view/component/page/404/index.js [hunk 19]
```diff
@@ -1,19 +1,21 @@
...
    */
   constructor(parentRoot) {
-    this._root = parentRoot;
+    super(parentRoot);
     this._render();
   }
 
   /**
-   * Add to the parent element text with error message and hyperlink to the login page.
+   * Renders page of error 404.
```
Comment (dkalinin) [inline app/view/component/page/404/index.js]: Renderes 404 page.

**Comment by kuzmavictor** [comment]
@dkalinin fixed, PTAL.

File: .gitignore (hunk 1 above)
Comment (dkalinin) [reply .gitignore]: Not fixed.

File: app/service/request/index.js [hunk 20]
```diff
@@ -25,21 +19,41 @@ export class RequestServiceProvider {
...
+   * @see RegistrationService
+   * @see FileListService
    * @private
    */
   _provideService() {
-    return {
-      authorizationService: this._authorizationService,
-      registrationService: this._registrationService, fileListService: this._fileListService};
+    return {authorizationService: new AuthorizationService(),
+      registrationService: new RegistrationService(), fileListService: new FileListService()};
```
Comment (dkalinin) [inline app/service/request/index.js]: Please remove hash and use private fields instead.

File: app/service/store/index.js [hunk 21]
```diff
@@ -36,29 +37,35 @@ export class Store {
...
   /**
    * Calls a mutator if it exists.
    *
    * @param {String} name - a name of mutation
-   * @param {...*} params -  an object that contains specified property for current mutator
+   * @param {...*} params - an object that contains specified property for current mutator
    */
   commit(name, ...params) {
-    this.mutations[name](this.state, ...params);
+    const mutationName = this.mutations[name];
```
Comment (dkalinin) [inline app/service/store/index.js]: mutationFunction

File: app/service/store/index.js [hunk 22]
```diff
@@ -36,29 +37,35 @@ export class Store {
...
+    if (typeof mutationName !== 'undefined' || mutationName !== null) {
```
Comment (dkalinin) [inline app/service/store/index.js]: Check if mutationFunction is function.

File: app/service/store/mutation/index.js [hunk 23]
```diff
@@ -7,8 +7,12 @@ export const mutations = {
    * @param {Object} state - an object that contains state
    * @param {BigInteger} id - an id of file
    */
-  removeItem(state, id) {
-    state.files = state.files.filter((file) => file.id !== id);
+  removeItem: (state, id) => {
+    const files = state.files;
+    const removedElement = files.map((file) => file.id).indexOf(id);
+    files.splice(removedElement, 1);
```
Comment (dkalinin) [inline app/service/store/mutation/index.js]: Use JS Array filter()

**Pull request merged**